### PR TITLE
feat: live update `/queue`

### LINKED ISSUE #92: Live update queue command
Changes to the queue (either new tracks being added or a track that finished playing) should automatically update the queue response interaction instead of the user having to click the arrow buttons to refresh it.

Visually, the queue button components could use a little change:
* Disable buttons when it's not possible to navigate to a specific page
* Use alt-codes instead of emojis
* Use primary blurple color instead of grey

## Current
image

## Expected
image

### PR DESCRIPTION
**Pull request by aquelemiguel**
Aims to close #92 and #94.

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -1,11 +1,22 @@
...
     client: serenity::Client,
 }
 
+pub struct GuildQueueInteractions;
+type QueueMessage = (Message, Arc<RwLock<usize>>);
+
+impl TypeMapKey for GuildQueueInteractions {
+    type Value = HashMap<GuildId, Vec<QueueMessage>>;
+}
+
```
Comment (joao-conde) [inline src/client.rs:19]: Move to its own module. In fact, wait for the approval and merge of https://github.com/aquelemiguel/parrot/pull/98 and let's merge this with the guild settings we have (possibly even rename to `GuildCache`)

File: src/commands/clear.rs [hunk 2]
```diff
@@ -31,5 +32,8 @@ pub async fn clear(
         v.drain(1..);
     });
 
+    drop(handler);
+    update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await;
+
     create_response(&ctx.http, interaction, "Cleared!").await
```
Comment (joao-conde) [inline src/commands/clear.rs:38]: Just an idea, maybe send the "Clear!" message first and then update the queue

File: src/commands/play.rs [hunk 3]
```diff
@@ -64,6 +65,8 @@ pub async fn _play(
         EnqueueType::PLAYLIST => enqueue_playlist(&call, url).await,
     };
 
+    update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await;
```
Comment (joao-conde) [inline src/commands/play.rs:68]: Again, I would update the queue at the end of this command, I think feedback to the command is important first.

File: src/commands/play.rs [hunk 4]
```diff
@@ -116,6 +119,12 @@ pub async fn _play(
 
 async fn calculate_time_until_play(queue: &[TrackHandle], flag: &PlayFlag) -> Option<Duration> {
     if !queue.is_empty() {
+        let non_livestreams = queue.iter().filter_map(|track| track.metadata().duration);
+
+        if non_livestreams.count() < queue.len() {
+            return Some(Duration::MAX);
+        }
```
Comment (joao-conde) [inline src/commands/play.rs:126]: This is not true for `/playtop` right?

File: src/commands/queue.rs [hunk 5]
```diff
@@ -1,25 +1,34 @@
...
         },
     },
-    prelude::SerenityError,
+    prelude::{RwLock, SerenityError, TypeMap},
 };
-use songbird::tracks::TrackHandle;
+use songbird::{tracks::TrackHandle, Event, TrackEvent};
 
 const EMBED_PAGE_SIZE: usize = 6;
+const COLLECTION_TIMEOUT: u64 = 3600;
```
Comment (joao-conde) [inline src/commands/queue.rs:31]: When this reaches 0 and it timeouts, does the embed disappear, like do we remove the embed message? If so, I would prefer `EMBED_TIMEOUT` and maybe with a little docstring specifying the time units (I know its seconds but still)

File: src/commands/queue.rs [hunk 6]
```diff
@@ -48,45 +55,43 @@ pub async fn queue(
...
-                        })
+                        .add_embed(create_queue_embed(&tracks, 0))
+                        .components(|components| build_nav_btns(components, 0, num_pages))
                 })
         })
         .await?;
 
-    let message = interaction.get_interaction_response(&ctx.http).await?;
+    let mut message = interaction.get_interaction_response(&ctx.http).await?;
+    let page_lock: Arc<RwLock<usize>> = Arc::new(RwLock::new(0));
```
Comment (joao-conde) [inline src/commands/queue.rs:68]: In reality, this is the `page`. Only when you call `page.read()` or `page.write()` do you get a lock. In the former a `page_rlock` in the latter a `page_wlock`

File: src/commands/queue.rs [hunk 7]
```diff
@@ -97,33 +102,43 @@ pub async fn queue(
         drop(handler);
 
         let num_pages = calculate_num_pages(&tracks);
+        let mut page = page_lock.write().await;
```
Comment (joao-conde) [inline src/commands/queue.rs:105]: And this would be

```rust
let mut page_wlock = page.write().await;
```

File: src/commands/queue.rs [hunk 8]
```diff
@@ -141,13 +156,36 @@ fn create_queue_embed<'a>(
...
+fn build_single_nav_btn(label: &str, is_disabled: bool) -> CreateButton {
+    let mut button = CreateButton::default();
+
+    button
+        .custom_id(label.to_string().to_ascii_lowercase())
+        .label(label)
+        .style(ButtonStyle::Primary)
+        .disabled(is_disabled);
+
+    button
```
Comment (joao-conde) [inline src/commands/queue.rs:173]: Is this not possible?

```suggestion
    CreateButton::default()
        .custom_id(label.to_string().to_ascii_lowercase())
        .label(label)
        .style(ButtonStyle::Primary)
        .disabled(is_disabled)
```

File: src/commands/queue.rs [hunk 9]
```diff
@@ -141,13 +156,36 @@ fn create_queue_embed<'a>(
...
+}
+
+pub fn build_nav_btns(
+    components: &mut CreateComponents,
+    page: usize,
+    num_pages: usize,
+) -> &mut CreateComponents {
+    components.create_action_row(|action_row| {
+        let (cant_left, cant_right) = (page < 1, page >= num_pages - 1);
+
```
Comment (joao-conde) [inline src/commands/queue.rs:183]: Nice. Except for the triggering NEW LINE *triggered*

File: src/commands/remove.rs [hunk 10]
```diff
@@ -49,6 +50,10 @@ pub async fn remove(
         handler.queue().modify_queue(|v| {
             v.remove(remove_index);
         });
+
+        drop(handler);
+        update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await;
```
Comment (joao-conde) [inline src/commands/remove.rs:55]: Again, maybe first send the feedback to the user ("Removed!") and then update the queue embed

File: src/commands/shuffle.rs [hunk 11]
```diff
@@ -26,6 +27,9 @@ pub async fn shuffle(
         )
     });
 
+    drop(handler);
+    update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await;
```
Comment (joao-conde) [inline src/commands/shuffle.rs:31]: Same idea, feedback to the user first.

File: src/commands/stop.rs [hunk 12]
```diff
@@ -28,5 +29,8 @@ pub async fn stop(
     }
 
     queue.stop();
-    return create_response(&ctx.http, interaction, "Stopped!").await;
+    drop(handler);
+
+    update_queue_messages(&ctx.http, &ctx.data, &call, guild_id).await;
+    create_response(&ctx.http, interaction, "Stopped!").await
```
Comment (joao-conde) [inline src/commands/stop.rs:35]: Feedback first

File: src/events/modify_queue_handler.rs [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    pub http: Arc<Http>,
+    pub ctx_data: Arc<RwLock<TypeMap>>,
+    pub call: Arc<Mutex<Call>>,
+    pub guild_id: GuildId,
+}
+
+#[async_trait]
+impl EventHandler for ModifyQueueHandler {
+    async fn act(&self, ctx: &EventContext<'_>) -> Option<Event> {
+        if let EventContext::Track(_) = ctx {
```
Comment (joao-conde) [inline src/events/modify_queue_handler.rs:28]: In reality, you only use this handler for the `Event::Track(TrackEvent::End)` so you know the `EventContext` will be of type Track. Since you don't use anything, I think you can remove the `if let`

File: src/events/modify_queue_handler.rs [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    for (message, page_lock) in messages.iter_mut() {
+        let handler = call.lock().await;
+        let tracks = handler.queue().current_queue();
+        drop(handler);
+
+        // has the page size shrunk?
+        let num_pages = calculate_num_pages(&tracks);
+
+        let mut page = page_lock.write().await;
+        *page = usize::min(*page, num_pages - 1);
```
Comment (joao-conde) [inline src/events/modify_queue_handler.rs:60]: This is all to compute size shrunk, just put it together, why the new line

File: src/tests/utils.rs [hunk 15]
```diff
@@ -1,29 +1,19 @@
 use serenity::model::prelude::User;
```
Comment (joao-conde) [inline src/tests/utils.rs:1]: Weird that clippy does not complain, but I think you do not need this import

File: src/tests/utils.rs [hunk 16]
```diff
@@ -1,29 +1,19 @@
...
-    user.name = "hello world".to_string();
-    user.discriminator = 1234;
-
-    let result = get_full_username(&user);
-    assert_eq!(result, "hello world#1234");
-}
+use crate::utils::get_human_readable_timestamp;
 
 #[test]
 fn test_get_human_readable_timestamp() {
```
Comment (joao-conde) [inline src/tests/utils.rs:7]: Add a test case where it receives a `None`

File: src/utils.rs [hunk 17]
```diff
@@ -28,19 +25,23 @@ pub async fn create_response(
...
+
+        if hours < 1 {
+            format!("{:02}:{:02}", minutes, seconds)
+        } else {
+            format!("{}:{:02}:{:02}", hours, minutes, seconds)
+        }
     } else {
-        format!("{}:{:02}:{:02}", hours, minutes, seconds)
+        "∞".to_string()
     }
```
Comment (joao-conde) [inline src/utils.rs:45]: This can better be rewritten as a match now that we have three cases: Some duration and infinite, some duration and finite, and the rest

**Review by joao-conde — CHANGES_REQUESTED** [review]

File: src/commands/play.rs (hunk 4 above)
Comment (aquelemiguel) [reply src/commands/play.rs:126]: Yeah, you're right. Oversight by my part, it'll show "Estimated time until play: ∞" which is incorrect.

File: src/commands/queue.rs (hunk 8 above)
Comment (aquelemiguel) [reply src/commands/queue.rs:173]: Tried it as well, it returns a `&mut CreateButton`. Could slap it a `.to_owned` but I dunno.

File: src/commands/queue.rs (hunk 9 above)
Comment (aquelemiguel) [reply src/commands/queue.rs:183]: Looks a lot worse on GitHub than VSCode, lol.

File: src/commands/queue.rs (hunk 5 above)
Comment (aquelemiguel) [reply src/commands/queue.rs:31]: The embed gets edited with `QUEUE_EXPIRED`, seen here:

https://github.com/aquelemiguel/parrot/blob/cd54d3432bf957022736a58007cf4bf20a9ea0d3/src/commands/queue.rs#L125-L135

Comment (joao-conde) [reply src/commands/queue.rs:31]: Then I would say just to use `EMBED_TIMEOUT`

File: src/commands/queue.rs (hunk 8 above)
Comment (joao-conde) [reply src/commands/queue.rs:173]: I like that

**Comment by joao-conde** [comment]
Haven't forgotten but will wait for merge conflicts fix 👍

File: src/commands/shuffle.rs [hunk 18]
```diff
@@ -26,7 +27,11 @@ pub async fn shuffle(
         )
     });
 
-    create_response(&ctx.http, interaction, "Shuffled successfully!").await
+    drop(handler);
+
+    create_response(&ctx.http, interaction, "Shuffled successfully!").await?;
```
Comment (afonsojramos) [inline src/commands/shuffle.rs:32]: Can you move the text to `strings.rs`?

File: src/commands/stop.rs [hunk 19]
```diff
@@ -28,5 +29,9 @@ pub async fn stop(
     }
 
     queue.stop();
-    return create_response(&ctx.http, interaction, "Stopped!").await;
+    drop(handler);
+
+    create_response(&ctx.http, interaction, "Stopped!").await?;
```
Comment (afonsojramos) [inline src/commands/stop.rs:34]: Can you move the text to `strings.rs`?

File: src/commands/clear.rs [hunk 20]
```diff
@@ -31,5 +32,9 @@ pub async fn clear(
         v.drain(1..);
     });
 
-    create_response(&ctx.http, interaction, "Cleared!").await
+    drop(handler);
+
+    create_response(&ctx.http, interaction, "Cleared!").await?;
```
Comment (afonsojramos) [inline src/commands/clear.rs:37]: Can you move the text to `strings.rs`?

**Review by afonsojramos — CHANGES_REQUESTED** [review]

File: src/utils.rs [hunk 21]
```diff
@@ -39,6 +37,51 @@ pub async fn create_embed_response(
...
+
+    let footer_text = format!(
+        "Track duration: {}\nEstimated time until play: {}",
+        get_human_readable_timestamp(metadata.duration),
+        get_human_readable_timestamp(Some(estimated_time))
+    );
+
+    embed.footer(|footer| footer.text(footer_text));
+    embed
+}
```
Comment (afonsojramos) [inline src/utils.rs:83]: This function should be deleted as it is now in `play.rs`.

**Review by afonsojramos — CHANGES_REQUESTED** [review]

**Review by joao-conde — APPROVED** [review]

**Review by afonsojramos — APPROVED** [review]
